backfill_rpc/src/rpc.rs

```diff
@@ -173,53 +201,194 @@ impl FinalizedSlotGetter for BackfillRPC {
+                        let retry = s.contains("429 Too Many Requests");
+                        if retry {
+                            self.is_in_backoff.store(true, Ordering::Relaxed);
+                        }
+                        retry
+                    },
+                    _ => false,
+                }
+            })
+            .await;
+        self.backoff_finish_notify.notify_one();
+        result
+    }
+}
+
+impl Deref for BackoffRpcClient {
+    type Target = RpcClient;
+
+    fn deref(&self) -> &Self::Target {
+        &self.client
+    }
+}
+
+#[cfg(all(test, feature = "rpc_tests"))]
+mod tests {
+    use solana_client::{
+        client_error::reqwest,
+        rpc_request::RpcRequest,
+        rpc_sender::{RpcSender, RpcTransportStats},
+    };
+    use warp::Filter;
+
+    use super::*;
+
+    struct TooManyRequestsRpcSender(String);
+
+    #[async_trait]
+    impl RpcSender for TooManyRequestsRpcSender {
+        async fn send(
+            &self,
+            request: RpcRequest,
+            _params: serde_json::Value,
+        ) -> Result<serde_json::Value, ClientError> {
+            let err = reqwest::get(format!("http://{}/", self.0))
+                .await
+                .expect("request to test server to succeed");
+            Err(ClientError::new_with_request(
+                ClientErrorKind::Reqwest(
+                    err.error_for_status().expect_err("429 server must indeed return 429"),
+                ),
+                request,
+            ))
+        }
+        fn get_transport_stats(&self) -> RpcTransportStats {
+            Default::default()
+        }
+        fn url(&self) -> String {
+            Default::default()
+        }
+    }
+
+    #[tokio::test]
+    async fn test_rpc_get_signatures_by_address() {
+        let client = BackfillRPC::connect("https://api.mainnet-beta.solana.com".to_string());
+        let signatures = client
+            .get_signatures_by_address(
+                Some(Signature::default()),
+                None,
+                &Pubkey::from_str("Vote111111111111111111111111111111111111111").unwrap(),
+            )
+            .await
+            .unwrap();
+
+        assert_eq!(signatures.len(), 1000)
+    }
+
+    #[tokio::test]
+    async fn test_backoff_client_failure() {
+        // Create a route that matches requests to "/" (the root path).
+        let route = warp::path::end().map(|| {
+            warp::reply::with_status(
+                warp::reply::json(&serde_json::Value::Null),
+                warp::http::StatusCode::TOO_MANY_REQUESTS,
+            )
+        });
+
+        let (addr, server) = warp::serve(route).bind_ephemeral(([127, 0, 0, 1], 0));
+        let _server_handle = tokio::spawn(server);
+
+        let client =
+            RpcClient::new_sender(TooManyRequestsRpcSender(addr.to_string()), Default::default());
+
+        let client = BackfillRPC { client: BackoffRpcClient::new_test(client) };
```

Comment:
```suggestion
        let client = BackfillRPC {
            client: {
                let backoff = ExponentialBackoff::from_millis(1).factor(1000);
                BackoffRpcClient {
                    client: Arc::new(client),
                    backoff,
                    is_in_backoff: Arc::new(AtomicBool::new(false)),
                    backoff_finish_notify: Arc::new(Notify::new()),
                }
            },
        };
```

Reply:
instead of using `new_test` method as a part of the trait for test purposes only

Reply:
If it will not be reusable it should be enough. Although if we plan doing this occasionally, then the trait might be ok

Reply:
it's not in the trait really, only in the struct impl, and i've pushed a test that uses the `new_test` function too, so i'll push for keeping it